### ISSUE TITLE
about batch size of the sgd algorithm

### ISSUE THREAD

**Issue by HeGaoYuan** [dennybritz/rnn-tutorial-gru-lstm#4]
Firstly, thank you very much! Your blog helps so many people to learn the RNN.
I have some questions about the parameter of batch. I am a newer to deep learning, if my question looks stupid, please forgive me. 
I have learned that when we use the sgd algorithm to optimize the loss function of CNN, we always give sgd  a batch size, but I never use the batch size equals to 1, I think one is too small.
Because I think when batch size equals to 1, the below equation  is not right.(the screenshot is from the website book neural networks and deep learning, [http://neuralnetworksanddeeplearning.com/chap1.html])

<img width="655" alt="2016-05-24 12 41 44" src="https://cloud.githubusercontent.com/assets/16939557/15492620/dcaf95a2-21af-11e6-9a31-80625865ed4a.png">

But I have read your blog and github code, I found that both in the RNN and LSTM, you both use the batch size equals to 1. **So my first question is why you use the batch size equals to 1?**
And I found that your code do not support to change the batch size of the sgd algorithm. **I am trying to modify your code to support change the batch size. Or do you think it is necessary to modify it?**

**Comment by dennybritz** [comment]
> So my first question is why you use the batch size equals to 1.

There is no specific reason. I didn't want to make the code complicated by adding an extra batch dimension and padding the sentences to the same length. When you use batches you need to make sure that all inputs are of the same size (sentence length) by adding special padding tokens. The code was written for learning/demonstration purposes, so I wanted to keep it as simple as possible. 

>  I am trying to modify your code to support change the batch size. Or do you think it is necessary to modify it?

Yes and no. If it's for your own learning purposes I think it's a good exercise to modify the code to support batches. However, if you want to use the code in some kind of production system or to train on a large dataset I'd recommend against it. The code here isn't optimized at all, you'd be much better of using built-in models from Keras or Tensorflow for serious applications. They'll be much faster.

> I have learned that when we use the sgd algorithm to optimize the loss function of CNN, we always give sgd a batch size, but I never use the batch size equals to 1

SGD with batch size 1 works just fine. In fact, you can prove that it converges to the right solution. The reason we use batches is more practical: It's much faster in practice and can make better use of computational resources like GPUs that are good at large matrix multiplications.

**Comment by HeGaoYuan** [comment]
Your reply is so great!!!
Thank you so much, it is very helpful.
